minimal_memory_lock/minimal_memory_lock.cpp

```diff
@@ -0,0 +1,273 @@
+    return;
+  }
+
+  ret = pthread_attr_setstacksize(&attr, stacksize);
+  if (ret != 0) {
+    std::cerr << "Failed to call pthread_attr_setstacksize. Error code: " <<
+      strerror(errno) << std::endl;
+    return;
+  }
+  ret = pthread_setattr_default_np(&attr);
+  if (ret != 0) {
+    std::cerr << "Failed to call pthread_setattr_default_np. Error code: " <<
+      strerror(errno) << std::endl;
+    return;
+  }
+}
+
+}  // namespace
+
+class MinimalPublisher : public rclcpp::Node
+{
+public:
+  MinimalPublisher()
+  : Node("minimal_publisher"), count_(0)
+  {
+    publisher_ = this->create_publisher<std_msgs::msg::String>("topic", 10);
+    auto timer_callback =
+      [this]() -> void {
+        auto message = std_msgs::msg::String();
+        message.data = "Hello, world! " + std::to_string(this->count_++);
+        this->publisher_->publish(message);
+        auto [minor, major] = get_new_page_faults();
+        RCLCPP_INFO(
+          this->get_logger(), "Page faults during spin: [minor: %d, major: %d]", minor, major);
+      };
+    timer_ = this->create_wall_timer(500ms, timer_callback);
+  }
+
+private:
+  rclcpp::TimerBase::SharedPtr timer_;
+  rclcpp::Publisher<std_msgs::msg::String>::SharedPtr publisher_;
+  size_t count_;
+};
+
+int main(int argc, char * argv[])
+{
+  // Force flush of the stdout buffer.
+  setvbuf(stdout, NULL, _IONBF, BUFSIZ);
+
+  // Argument count and usage
+  if (rcutils_cli_option_exist(argv, argv + argc, "-h")) {
+    print_usage();
+    return 0;
+  }
+
+  // Configuration variables
+  size_t max_process_memory = 0;
+  size_t stack_size = 0;
+  bool lock_memory = false;
+  bool use_sleep = false;
+
+  // Optional argument parsing
+  if (rcutils_cli_option_exist(argv, argv + argc, OPTION_LOCK_MEMORY)) {
+    lock_memory = true;
+  }
+  if (rcutils_cli_option_exist(argv, argv + argc, OPTION_MEMORY_PREALLOCATION)) {
+    max_process_memory = 1024 * 1024 *
+      std::stoul(rcutils_cli_get_option(argv, argv + argc, OPTION_MEMORY_PREALLOCATION));
+  }
+  if (rcutils_cli_option_exist(argv, argv + argc, OPTION_THREAD_STACK_SIZE)) {
+    stack_size = 1024 *
+      std::stoul(rcutils_cli_get_option(argv, argv + argc, OPTION_THREAD_STACK_SIZE));
+  }
+  if (rcutils_cli_option_exist(argv, argv + argc, OPTION_USE_SLEEP)) {
+    use_sleep = true;
+  }
+
+  // Change a new default thread stack size to reduce memory foot-print
+  if (stack_size > 0) {
+    set_default_thread_stacksize(stack_size);
+  }
+
+  print_process_memory("Process memory before node creation: ");
+
+  // Create nodes
+  rclcpp::init(argc, argv);
+  auto node = std::make_shared<MinimalPublisher>();
+
+  // Sleep required to avoid some initial minor page faults
+  // TODO(carlos): find out why we see minor page faults initially
+  // This could be caused by middleware threads being created concurrently
+  if (use_sleep) {
+    std::cout << "Sleeping here" << std::endl;
+    std::this_thread::sleep_for(500ms);
+  }
```

Comment:
This seems kind of sketchy, as I don't observe it outside of ROS. I do buy that the theory that background threads are starting.

That said, you can use `getrusage(RUSAGE_THREAD, &usage);` to get the current thread's page faults (I see your code is getting the page fault for the entire process). This way you won't get fooled by the page fault of the "background" threads. You can then eliminate the entire `--sleep` option.

Reply:
The problem here is that I want to make sure the DDS threads have been created. Otherwise, you will get page faults in the real-time code section. This is the main reason I prefer to keep `RUSAGE_SELF`. 

Since there is no rclcpp API to check if the middleware threads have been initialized I see this as a limitation. This is why I'm putting a sleep here with a comment, to highlight the problem in the hope we are able to overcome it in the future.

The same issue could cause lost packages if the middleware is not ready. See a related issue: https://github.com/ros2/rmw_connext/issues/112

Reply:
Aren't the DDS threads normal priority threads? Those page faults will occur in the normal priority threads, right? My understanding is that page faults in normal priority threads is fine (unless priority-inheritance mutex is involved).

Reply:
Ah that makes sense. I forgot that some of the ROS-related threads needs to be RT as well.